paper.tex

```diff
@@ -540,6 +536,8 @@ \subsection{Pipeline Design} \label{subsec:pipeline-design}
 
 All input segments first get converted to a cubic Bézier by degree elevation. Our flattening routine (see Algorithm~\ref{alg:flatten-cubic}) starts by computing the error estimate for geometric Hermite interpolation from the cubic Bézier to an Euler spiral segment. If the error exceeds the desired tolerance threshold, the cubic segment is iteratively subdivided in half and a new error estimate is computed. Once the error satisfies the tolerance, the Euler spiral segment is directly flattened to a polyline by invoking a subroutine that is parameterized by the stroke half-width (see \texttt{es\_seg\_flatten\_offset} in Algoritm~\ref{alg:flatten-cubic}). For a regular fill, this parameter is set to $0$. For strokes, the sign of the parameter determines the position of the parallel curve relative to the source segment as well as the winding direction of the output lines.
 
+The adaptive subdivision in lowering cubic Béziers to Euler spiral segments is traditionally accomplished through recursion; in the case of subdivision there is a recursive call for each of the two subdivisions of the range. As the execution model of compute shaders does not support recursion, we \emph{unroll} recursion into an iterative loop. The cookbook technique for such unrolling is to store the stack explicitly as an array, which is potentially expensive (it increases pressure on registers). We exploit the specific nature of this recursion, encoding the entire state of the stack into two scalar values: the size of the range $dt$, and the scaled start of the range $t0\_u$, so that the range is $t0\_u \cdot dt .. (t0\_u + 1) \cdot dt$. Initial values are $1$ and $0$ respectively. Pushing the call stack is represented by halving $dt$ and doubling $t0\_u$, which preserves the start of the range but halves the size. Accepting an approximation (a leaf in the call stack) is represented by incrementing $t0\_u$. At this point, the decision of whether to pop the stack or continue is dependent on whether $t0\_u$ is even or odd, respectively. If even, popping the stack is accomplished by doubling $dt$ and halving $t0\_u$, and this is repeated until $t0\_u$ becomes odd. The repeated stack popping can be accomplished without iteraction by using the ``count trailing zeros'' intrinsic to determine the number of stack pops $n$, then multiplying $dt$ by $2^n$, and shifting $t0\_u$ right by $n$ bits.
+
```

Comment:
Optional: may want to add  `(see Algorithm~\ref{alg:flatten-cubic})` at the end of the paragraph.